### PR TITLE
LIME-1045 Reduce and optimize SSM parameter reads within FraudAPI

### PR DESCRIPTION
**Pull request by smsgds**
## Proposed changes

~~Rebased off changes from un-merged LIME-952 (will rebase this after merge)~~
Rebased to main
Pre-merge checks depend on - stub pr - https://github.com/govuk-one-login/ipv-third-party-stubs/pull/191

	- Parameters: Switch to reading multiple parameters from a path, inplace of indivudal reads.
	- Parameters: Add SSMProvider and SecretsProvider maxCache age, random offset added to avoid the powertools cache expiring at the same time for muliple lambdas.
	- Parameters: Suppress sonar warning for ThreadLocalRandom in getSSMProvider() and getSecretsProvider(), as it is not used cryptographically.
	- Parameters: Note behaviour where parameters all read at startup has not changed. (work is done that changing this should now be trivial if needed)

	- AC Tests: Add missing API based HappyPath test for Kenneth
	- AC Tests: Add step to above to confirm correct test user was used.
	- AC Tests: Fix VC being retrived multiple times per test.
	- AC Tests: Fix Firefox headless incorrect argument
	- AC Tests: Add Browser choice to run-local-test.sh (default chrome-headless)
	- AC Tests: Fix FraudAPIPage Logger calling internal Java logging, instead of log4j.
	- AC Tests: Add ablity for API test to test fraud check responses.
	- AC Tests: Add API test for FWAIT_OUTOFTIME and assert the response matches common-express to forward as oauth.
	- AC Tests: Add API test for overall HappyPath, FWAIT_INTIME, PWAIT_INTIME, PWAIT_OUTOFTIME, ALL_WAIT_INTIME
	- AC Tests: Add edge case test for scenario AHS_PEP_TECH_FAIL and assert evidence checks match what is expected

	- Lib: Add ClientFactoryService from DL/Passport
	- Lib: Add DynamoDbEnhancedClient to ClientFactoryService
	- Lib: Update ClientFactoryService to set awsRegion, sdkHttpClient and environmentVariableCredentialsProvider on each client - fixes exceptions due to no credentialsProvider and centralizes the region enforcement.
	- Lib: Add support for creating a HTTPConnectionKeepAliveStrategy in ClientFactoryService.
	- Lib: Add ParameterStoreService from DL/Passport
	- Lib: Add ParameterPrefix Enum and rework ParameterStoreService to avoid needing to add new methods for every prefix combination.
	- Lib: Add ParameterStoreParameters to centralize parameter strings constants.
	- Lib: Moved HTTPReplyHelper, HttpRetryer, Sleephelper, HttpRetryStatusConfig into lib to be consistent with DL/Passport.
	- Lib: Moved HttpResponseFixtures in the the lib equivilent.
	- Lib: Add support for headerIterator() in HttpResponseFixtures.
	- Lib: Add KeyCertHelper from DL/Passport
	- Lib: Moved ServiceFactory here
	- Lib: ServiceFactory remove FraudCheck only services.
	- Lib: Add ResultItemStorageService for managing fraud check result items.
	- Lib: Add CommonExpressOAuthError to help set oauth errors.
	- Lib: Add KeepAlive (TODO)
	- Lib: Add CheckType ACTIVITY_HISTORY_CHECK to enable the check to be handled the same way as the others.
	- Lib: Change Check constructor to require the CheckType enum to prevent creating invalid check objects with fraudCheck set to arbitary strings.
	- Lib: Update EvidenceHelper to use the CheckType in determining unique handling of each check in checkDetails/failedCheckdetails
	- Lib: Update EvidenceHelper to process ACTIVITY_HISTORY_CHECK only when score not 0. (to remain consistent with prior behaviour)
	- Lib: Removed KeyCert support utils and test fixtures.

	- FraudCheck: Align handler setup with DL/Passport. (ServiceFactory, ClientService Factory, Verification Service etc)
	- FraudCheck: Correct handler server error responses to respond with the corect server (and not rely on the front end to suppress error messages)
	- FraudCheck: Allow the previous error message responses via the dev env feature toggle DEV_ENVIRONMENT_ONLY_ENHANCED_DEBUG.
	- FraudCheck: Switch to ResultItemStorageService for FraudResultItem storage.
	- FraudCheck: Added tracking of function init duration (as metric) and log the lambda overall life duration at the start of each journey.
	- FraudCheck: Add ThirdPartyAPIServiceFactory to centralize setup of and provide the three services (Fraud,Pep, Token)
	- FraudCheck: Removed ContraindicationDynamoDBMapper unused ContraindicationDynamoDBMapper
	- FraudCheck: Removed ContraindicationMapper interface.
	- FraudCheck: Renamed ContraIndicatorRemoteMapper to ContraindicationMapper.
	- FraudCheck: Add ThirdPartyCloseableHttpClientFactory to create the HTTP client.
	- FraudCheck: Correctly mock person identity service.
	- FraudCheck: Throw OAuthErrorResponseException when failing to retrieve person identity.
	- FraudCheck: Ensure all current errors are returned as OAuth2Error.SERVER_ERROR.
	- FraudCheck: Remove CC1 parameters reading
	- FraudCheck: Remove pepEnabled feature toggle (now always on)
	- FraudCheck: Capture ACTIVITY_HISTORY_CHECK when activityHistoryScore is greater than 0 (tbc paths for 0)
	- FraudCheck: Removed older keystore http client support methods

	- IssueCred : Align handler setup with DL/Passport. (ServiceFactory, VerifiableCredentialService, KMSSigner etc)
	- IssueCred : Remove IssueCredentialConfigurationService
	- IssueCred : Align VerifiableCredentialService with DL/Passport.
	- IssueCred : Switch to ResultItemStorageService for FraudResultItem retrieval.

	- Actions   : Removed build.yml - sonar scan already done as part of pre-merge-check.yml
	- Unittests : coverage now 80%.
	- Sonar     : Fix several code smells that got noticed during refactoring.

----
Follow up changes -

	- AC Tests: run-local-test.sh browser selection using wrong variable
	- AC Tests: fix driverPool.get().close() crashing firefox driver
	- Template. : mirror keystore changes in LIME-592 to reflect in this pr.
	- FraudCheck: mirror keystore changes in LIME-592 to reflect in this pr.

	- Lib       : Move createHTTPConnectionKeepAliveStrategy out of ClientFactoryService to avoid issuecred needing the extra dependency
	- Lib       : Removed KeyCert support utils and test fixtures.

	- FraudCheck: Fix code smell due to secretsProvider no longer being used.
	- FraudCheck: Removed older keystore http client support methods
	- FraudCheck: Add missing lambda lifetime logging and add all code after the Initiating lambda log. 
	- FraudCheck: Fix calling getAllParametersFromPathWithDecryption when getAllParametersFromPath is sufficient

	- IssueCred : Add missing lambda lifetime logging and add all code after the Initiating lambda log.

	- Pact      : Align with changes in the handler/service factory
	- Pact      : Align with correction to activity history check and implementation
	- Pact      : Add slf4j_log4j12_version "2.0.12" so contract test debug can be requested
	- Pact      : add boolean ENABLE_FULL_DEBUG (default off) to setup debug logging from pact

- [LIME-1045](https://govukverify.atlassian.net/browse/LIME-1045)

[LIME-1045]: https://govukverify.atlassian.net/browse/LIME-1045?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/fraudcheck/src/main/java/uk/gov/di/ipv/cri/fraud/api/service/IdentityVerificationService.java [hunk 1]
```diff
@@ -199,15 +200,15 @@ public IdentityVerificationResult fraudCheckStep(PersonIdentity personIdentity,
...
+            fraudCheckResult = thirdPartyFraudGateway.performFraudCheck(personIdentity, token);
         } catch (Exception e) {
             LOGGER.error(ERROR_MSG_CONTEXT, e);
             eventProbe.counterMetric(FRAUD_CHECK_REQUEST_FAILED);
 
             String errorMessage = e.getMessage();
 
-            if (e instanceof OAuthErrorResponseException) {
-                errorMessage = ((OAuthErrorResponseException) e).getErrorReason();
+            if (e instanceof OAuthErrorResponseException oAuthErrorResponseException) {
```
Comment (gds-jbentham) [inline lambdas/fraudcheck/src/main/java/uk/gov/di/ipv/cri/fraud/api/service/IdentityVerificationService.java:210]: Is this double entry a mistake?

File: lib/src/main/java/uk/gov/di/ipv/cri/fraud/library/util/KeyCertHelper.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package uk.gov.di.ipv.cri.fraud.library.util;
```
Comment (gds-jbentham) [inline lib/src/main/java/uk/gov/di/ipv/cri/fraud/library/util/KeyCertHelper.java:1]: Is this file still needed?

File: lib/src/testFixtures/java/uk/gov/di/ipv/cri/fraud/library/util/CertAndKeyTestFixtures.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package uk.gov.di.ipv.cri.fraud.library.util;
```
Comment (gds-jbentham) [inline lib/src/testFixtures/java/uk/gov/di/ipv/cri/fraud/library/util/CertAndKeyTestFixtures.java:1]: Same here is this still needed

File: lambdas/fraudcheck/src/main/java/uk/gov/di/ipv/cri/fraud/api/service/IdentityVerificationService.java (hunk 1 above)
Comment (smsgds) [reply lambdas/fraudcheck/src/main/java/uk/gov/di/ipv/cri/fraud/api/service/IdentityVerificationService.java:210]: No this is the new format to avoid a cast of the exception when its is known by the jvm due to instanceof being called.

https://rules.sonarsource.com/java/tag/java16/RSPEC-6201/

File: lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/fraud/api/util/EvidenceHelper.java [hunk 4]
```diff
@@ -29,54 +27,61 @@ public static Evidence fraudCheckResultItemToEvidence(
...
+        Check check = new Check(checkType);
 
-        if (activityHistoryEnabled) {
-            if (resultItem.getActivityHistoryScore() != null
-                    && resultItem.getActivityHistoryScore() != 0) {
-                Check check = new Check();
-                check.setActivityFrom(resultItem.getActivityFrom());
-                check.setIdentityCheckPolicy("none");
-                checkList.add(check);
+        switch (checkType) {
```
Comment (ElliotMurphyGDS) [inline lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/fraud/api/util/EvidenceHelper.java:66]: Nice work here

File: lib/src/main/java/uk/gov/di/ipv/cri/fraud/library/service/ServiceFactory.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        this.eventProbe = new EventProbe();
+
+        this.clientFactoryService = new ClientFactoryService();
+        this.parameterStoreService = new ParameterStoreService(clientFactoryService);
+
+        this.commonLibConfigurationService =
+                new uk.gov.di.ipv.cri.common.library.service.ConfigurationService();
+
+        this.sessionService = new SessionService();
+        this.auditService =
```
Comment (ElliotMurphyGDS) [inline lib/src/main/java/uk/gov/di/ipv/cri/fraud/library/service/ServiceFactory.java:47]: perhaps me being picky but can we move the creation of the auditService/commonlibconfiguration service into a "createAuditService" method like we had before? Just helps make the seperation of the services clearer for me

File: acceptance-tests/src/test/java/gov/di_ipv_fraud/utilities/Driver.java [hunk 6]
```diff
@@ -64,7 +65,7 @@ public static WebDriver get() {
                     WebDriverManager.firefoxdriver().setup();
 
                     FirefoxOptions firefoxOptions = new FirefoxOptions();
-                    firefoxOptions.addArguments("--headless");
+                    firefoxOptions.addArguments("-headless");
```
Comment (ElliotMurphyGDS) [inline acceptance-tests/src/test/java/gov/di_ipv_fraud/utilities/Driver.java:68]: We sure this is correct or is this a typo? Near sure its supposed to be --

**Review by gds-jbentham — APPROVED** [review]

**Review by ElliotMurphyGDS — APPROVED** [review]
